### DISCUSSION TITLE
useLoaderData vs useSuspenseQuery

### DISCUSSION THREAD

**Discussion by guy-ecton** [TanStack/router#1563] [Q&A]
I've set up a component loader with react-query (as shown in the documentation), but I "forgot" to change the component itself, and I kept it with `useLoaderData()`.
```typescript
loader: async ({ context: { quernClient } }) => queryClient.ensureQueryData(getPostsOptions)

// ....
function Comp() {
  const posts = useLoaderData()
}
```
To my surprise, it worked. `posts` contained the output I expected.

If my loader returns more than 1 key (like a promise and a primitive), using `useSuspenseQuery` in the component seems limiting, as it is tightly bound to react-query.

So, which should we use?
Thanks

**Comment by TkDodo** [comment]
with the shown code, you will only ever fetch data once and never re-validate, because that's what `ensureQueryData` does. Also, without `useQuery` or `useSuspenseQuery`, there is no query subscription, so your data is eligible for garbage collection.

The way I use it is with `useSuspenseQuery` and I don't `useLoaderData()` at all. In fact, I don't return anything from the loader so that it can't be used accidentally. For me, the loader is the place to seed the cache.

**Reply by guy-ecton** [reply]
This is (almost) same code example as in the documentation: https://tanstack.com/router/latest/docs/framework/react/guide/external-data-loading#a-more-realistic-example-using-tanstack-query

> For me, the loader is the place to seed the cache.

Oh, interesting. Now that I re-read the docs it indeed seems like the loader shouldn't return anything (although the fat arrow in `() => queryClient....` implies it does return something, as opposed to `() => { queryClient... }`).
So I think thats why `useLoaderData` did work as well.

If you have several calls (or several data pieces to seed) would you then use multiple `useSuspenseQuery`s?

**Reply by TkDodo** [reply]
> If you have several calls (or several data pieces to seed) would you then use multiple useSuspenseQuerys?

`useSuspenseQueries` is what I use.

**Reply by max-programming** [reply]
@TkDodo Right now the problem I am facing is that when I make the loader `async`, the `Suspense` component does not render any `fallback`, and it waits for the data to fetch like a normal `useLoaderData`. **Although I am using `useSuspenseQuery`**

It's fine if I don't want anything out of the `loader` but I need a `title` from the `loader` in the `head`

How can I achieve this?
Here's some code for reference

Non Async Loader - Works Great with `Suspense` and `useSuspenseQuery`
```ts
  loader({ context, params, deps }) {
    const { chatId } = params;
    context.queryClient.ensureQueryData(queries.chats.withMessages(chatId));
  },
  head() {
    return {
      meta: [
        {
          title: `New Chat | Chat Assistant`, // No dynamic title
        },
      ],
    };
  },
```

Async Loader - Does not show `Suspense` fallback and **acts like** `useLoaderData`
```ts
  async loader({ context, params, deps }) {
    const { chatId } = params;
    const { chat } = await context.queryClient.ensureQueryData(queries.chats.withMessages(chatId));
    return { title: chat.title }
  },
  head({ loaderData }) {
    return {
      meta: [
        {
          title: `${loaderData?.title ?? "New Chat"} | Chat Assistant`, // Dynamic title
        },
      ],
    };
  },
```

**Reply by conficiusa** [reply]
same issue . suspense fallback not rendering

**Reply by schiller-manuel** [reply]
you cannot have both. either block via awating in the loader, then you have loader data you can use in head
if you are not awaiting in the loader, you cannot access the loader data in head as head is built before the initial response is returned to the client, and your data streams in later.

**Reply by TkDodo** [reply]
@max-programming the original question as about `router`, but your code snippet indicates you’re using `start`, so there’s a difference. You need to await in the loader and `useLoaderData` if you want that information available in `head` afaik.

**Reply by max-programming** [reply]
@TkDodo I understand. But check out [this example](https://codesandbox.io/p/devbox/awaited-loader-suspense-8ckxvh)

You will notice clicking on "Deferred" link, it shows Suspense fallback properly
But upon clicking "Deferred (Failing)" link, it waits for the page to fully load and Suspense fallback is never shown

The only difference between the two routes is the failing one has `async await` on the `loader`
Is there a way to tackle this issue? That are able to use `await` along with suspense fallback

**Reply by TkDodo** [reply]
if you `await` in the loader, the router will show the nearest suspense boundary. Every route is wrapped in a suspense boundary _if_ if you add a `pendingComponent`. You have no `pendingComponent`, so there is nothing to show. We can’t show the `Suspense` fallback you have defined inside a component of a route because the component will only render after the loader has finished. So add:

```
pendingComponent: () => "Pending Middleman..."
```

or define a `defaultPendingComponent` on the router itself to get a fallback shown if you await in the loader. Additionally, if you have that and you remove your own suspense boundary, the `pendingComponent` would also be used if you don’t await in the loader, because what happens then is that the component renders and `useSuspenseQuery` will throw the promise to the nearest boundary, which again, is the `pendingComponent`.

**Reply by max-programming** [reply]
@TkDodo It makes sense now. I didn't know about `pendingComponent`. Thanks

**Reply by danielytics** [reply]
I'm looking at this discussion because I'm trying to figure out the correct way to use Tanstack Query with Tanstack Router.

In the [Tanstack Query documentation](https://tanstack.com/query/latest/docs/framework/react/guides/prefetching#router-integration) there's an example:

```typescript
  loader: async ({
    context: { queryClient },
    routeContext: { articleQueryOptions, commentsQueryOptions },
  }) => {
    // Fetch comments asap, but don't block
    queryClient.prefetchQuery(commentsQueryOptions)

    // Don't render the route at all until article has been fetched
    await queryClient.prefetchQuery(articleQueryOptions)
  },
```

It also links to [this example](https://tanstack.com/router/.latest/docs/framework/react/examples/basic-react-query-file-based) which uses `ensureQueryData`:

```typescript
import { createFileRoute } from '@tanstack/react-router'
import * as React from 'react'
import { Link, Outlet } from '@tanstack/react-router'
import { useSuspenseQuery } from '@tanstack/react-query'
import { postsQueryOptions } from '../postsQueryOptions'

export const Route = createFileRoute('/posts')({
  loader: ({ context: { queryClient } }) =>
    queryClient.ensureQueryData(postsQueryOptions),
  component: PostsLayoutComponent,
})

function PostsLayoutComponent() {
  const postsQuery = useSuspenseQuery(postsQueryOptions)
  const posts = postsQuery.data

  return (
    <div className="p-2 flex gap-2">
      <ul className="list-disc pl-4">
        {[...posts, { id: 'i-do-not-exist', title: 'Non-existent Post' }].map(
          (post) => {
            return (
              <li key={post.id} className="whitespace-nowrap">
                <Link
                  to="/posts/$postId"
                  params={{
                    postId: post.id,
                  }}
                  className="block py-1 text-blue-600 hover:opacity-75"
                  activeProps={{ className: 'font-bold underline' }}
                >
                  <div>{post.title.substring(0, 20)}</div>
                </Link>
              </li>
            )
          },
        )}
      </ul>
      <hr />
      <Outlet />
    </div>
  )
}
```

And the [Tanstack Router documentation](https://tanstack.com/router/latest/docs/framework/react/guide/deferred-data-loading#deferred-data-loading-with-external-libraries) it uses `prefetchQuery` to kick off async data and `ensureQueryData` to get the immediately needed data:

```typescript
  loader: async ({ context: { queryClient } }) => {
    // Kick off the fetching of some slower data, but do not await it
    queryClient.prefetchQuery(slowDataOptions())

    // Fetch and await some data that resolves quickly
    await queryClient.ensureQueryData(fastDataOptions())
  },
```

In all three examples, the loader data is never used, instead `useQuery` or `useSuspenseQuery` is used to actually obtain the data.

My question is: why is `ensureQueryData` ever used in the loader? Why populate the loader cache when its never accessed? In the last example, its thrown away anyway, so wouldn't `await prefetchQuery` have done the same thing? Why not just always use non-awaited `prefetchQuery` for data that can load asynchronously and awaited `prefetchQuery` for data that is needed immediately? 

The documentation isn't very clear, especially when the React Query docs talk about `prefetchQuery` but then link to an example that uses `ensureQueryData`. When would I prefer one way over the other?

Or is it really the same and I can use either one interchangeably?

Thanks for the help!

**Reply by coleea** [reply]
@danielytics As I understand, both `prefetchQuery`  and  `ensureQueryData` call  `fetchQuery` internally.
the main difference is that `prefetchQuery` always fetch data. `ensureQueryData` only fetch data when cache is not present. So `ensureQueryData` can prevent unnecessary data fetching (if the cache already exists).
But `prefetchQuery` is also useful if latest updated data is needed (that is not stale data).
Let me know If I am wrong

**Reply by TkDodo** [reply]
>  is that prefetchQuery always fetch data.

prefetchQuery respects staleTime, so it doesn’t always fetch. See also: 
- https://github.com/TanStack/query/discussions/9135

**Reply by hanneswidrig** [reply]
I have a question related to @danielytics above, the [router](https://tanstack.com/router/latest/docs/guide/external-data-loading#a-more-realistic-example-using-tanstack-query) docs show this example.

```tsx
// src/routes/posts.tsx
const postsQueryOptions = queryOptions({
  queryKey: ['posts'],
  queryFn: () => fetchPosts(),
})

export const Route = createFileRoute('/posts')({
  // Use the `loader` option to ensure that the data is loaded
  loader: () => queryClient.ensureQueryData(postsQueryOptions),
  component: () => {
    // Read the data from the cache and subscribe to updates
    const {
      data: { posts },
    } = useSuspenseQuery(postsQueryOptions)

    return (
      <div>
        {posts.map((post) => (
          <Post key={post.id} post={post} />
        ))}
      </div>
    )
  },
})
```

My question is, if I'm using `useSuspenseQuery` in the component, what reasons would I have to use `queryClient.ensureQueryData`, `queryClient.fetchQuery`, `queryClient.prefetchQuery` or even the upcoming `queryClient.query` in the `loader` function?

I'm wondering if maybe it's to prevent a round-trip network call assuming I fetch the data before the component loads on the client?

**Reply by TkDodo** [reply]
> My question is, if I'm using useSuspenseQuery in the component, what reasons would I have to use queryClient.ensureQueryData, queryClient.fetchQuery, queryClient.prefetchQuery or even the upcoming queryClient.query in the loader function?

The loader can run long _before_ the component renders. For example, with intent prefetching, TanStack Router would start running the loader as soon as the user hovers a link that would go to a specific route. It can also run before the code-split bundle gets downloaded, as components get automatically split out with `React.lazy`

With just `useSuspenseQuery`, you have to download the component bundle, render it, and then suspense would kick off the data fetching once `useSuspenseQuery` gets invoked.

**Reply by hanneswidrig** [reply]
@TkDodo Thanks so much for the explanation! It sounds like any reasonably quick request I will need for a given route should be _retrieved_ ahead of time in the loader.
